### ISSUE TITLE
Different evaluation results during tuning and after tuning?

### ISSUE THREAD

**Issue by Kirayue** [maxpumperla/hyperas#163]
I used hyperas to do experiments, I encountered a problem that 
I randomly split data five times, and tuned the model on each split.
I found that model.evaluate results are different except the first split. I print l2 value, (train_loss, train_acc), (val_loss, val_acc) and model each tuning round.
After tuning, I use best_model that hyperas returned to evaluate the same train/val data, the results is different from turning. The model's memory location are the same. I have no idea what's happening.
Thank you.

![image](https://user-images.githubusercontent.com/13309212/39116646-8dc6e9f6-4717-11e8-9593-eef3f4fe350b.png)

**Comment by maxpumperla** [comment]
how can I reproduce this @Kirayue ?

**Comment by Kirayue** [comment]
Hi, @maxpumperla 
You can use this code
`import sys
import numpy as np
import time
import keras
from hyperopt import Trials, STATUS_OK, tpe
from hyperas.distributions import choice, uniform, conditional, loguniform
from hyperas import optim
from keras.layers import Dense, Input, Dropout, LSTM, Activation, BatchNormalization, Conv1D, MaxPooling1D, Flatten, Subtract, Lambda, Add, Concatenate, GaussianNoise, LeakyReLU, PReLU, ThresholdedReLU
from keras.models import Model
from keras import regularizers, initializers
from keras import backend as K
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

def split(random_state):
    data = np.loadtxt('./australian.dat')
    x_train, x_test, y_train, y_test = train_test_split(data[:,:-1], data[:, -1], test_size=0.1, random_state=random_state, stratify=data[:, -1])
    return x_train, x_test, y_train, y_test

def data():
    x_train = np.load('tmp/x_train.npy')
    y_train = np.load('tmp/y_train.npy')
    x_val = np.load('tmp/x_val.npy')
    y_val = np.load('tmp/y_val.npy')
    return x_train, y_train, x_val, y_val

def create_model(x_train, y_train, x_val, y_val):
    n = 237
    init = 'lecun_normal'
    act = 'selu'

    l2 = {{loguniform(-9.2, -2.3)}}

    batch_size = 32
    epochs = 500

    X_input = Input((14, ))
    X = Dense(n, kernel_initializer=init, kernel_regularizer=regularizers.l2(l2))(X_input)
    X = Activation(act)(X)
    X = Dense(1, kernel_initializer=init, kernel_regularizer=regularizers.l2(l2))(X)
    X_outputs = Activation('sigmoid')(X)

    model = Model(inputs = X_input, outputs = X_outputs)
    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
    model.fit(x_train, y_train, batch_size=batch_size, verbose=0, epochs=epochs, validation_data=(x_val, y_val))
    print("L2: {}".format(l2))
    loss, score = model.evaluate(x_train, y_train, verbose=0)
    print(loss, score)
    loss, score = model.evaluate(x_val, y_val, verbose=0)
    print(loss, score)
    print(model)
    return {'loss': loss, 'status': STATUS_OK, 'model': model}

if __name__ == "__main__":
    times = 5
    train_scores = np.empty((times, 1))
    val_scores = np.empty((times, 1))
    test_scores = np.empty((times, 1))
    train_losses = np.empty((times, 1))
    val_losses = np.empty((times, 1))
    test_losses = np.empty((times, 1))
    for i in range(times):
        start = time.time()
        seed = i * 12345 + 98765
        print()
        print(i)
        print("Seed: {}".format(seed))
        _x_train, x_test, _y_train, y_test = split(seed)
        x_train, x_val, y_train, y_val = train_test_split(_x_train, _y_train, test_size=1/9, random_state=1220, stratify=_y_train)

        scaler = StandardScaler()
        x_train = scaler.fit_transform(x_train)
        x_val = scaler.transform(x_val)
        x_test = scaler.transform(x_test)

        np.save('tmp/x_train.npy', x_train)
        np.save('tmp/y_train.npy', y_train)
        np.save('tmp/x_val.npy', x_val)
        np.save('tmp/y_val.npy', y_val)
        best_run, model = optim.minimize( model=create_model,
                                          data=data,
                                          algo=tpe.suggest,
                                          max_evals=60,
                                          trials=Trials(),
                                          rseed=seed,
                                          verbose=False)
        print("Best performing model chosen hyper-parameters:")
        print(best_run)
        print(model)
        train_loss, train_score = model.evaluate(x_train, y_train, verbose=0)
        print(train_loss, train_score)
        val_loss, val_score = model.evaluate(x_val, y_val, verbose=0)
        print(val_loss, val_score)
        test_loss, test_score = model.evaluate(x_test, y_test, verbose=0)
        print(test_loss, test_score)
        train_scores[i] = train_score
        val_scores[i] = val_score
        test_scores[i] = test_score 
        train_losses[i] = train_loss
        val_losses[i] = val_loss
        test_losses[i] = test_loss
        K.clear_session()
        end = time.time()
        print(end - start)

    print()
    print(np.mean(train_scores))
    print(np.mean(val_scores))
    print(np.mean(test_scores))
    print(np.mean(train_losses))
    print(np.mean(val_losses))
    print(np.mean(test_losses))
`
and I use this [toy dataset](http://archive.ics.uci.edu/ml/machine-learning-databases/statlog/australian/australian.dat)

**Comment by Kirayue** [comment]
Hi, @maxpumperla 
Got it, thanks for your reply.
But I still get confused why I ran `optim.minimize`  5 times, but the data function was called once? In other words, I ran `optim.minimize`  5 times but the data for optimization did not change.

**Comment by maxpumperla** [comment]
the data function is called 5 times as well, but 5 times with the same values: the values provided in `data`:

```
def data():
    x_train = np.load('tmp/x_train.npy')
    y_train = np.load('tmp/y_train.npy')
    x_val = np.load('tmp/x_val.npy')
    y_val = np.load('tmp/y_val.npy')
    return x_train, y_train, x_val, y_val
```

it doesn't matter what you put into your loop:

```
    for i in range(times):
       ...
        _x_train, x_test, _y_train, y_test = split(seed)
        x_train, x_val, y_train, y_val = train_test_split(...)
       ...
```

what's passed to `optim.minimize` is always just `data`. In other words, you can't just override `x_train` from outside `data`.

**Comment by Kirayue** [comment]
I override x_train by overriding ` .npy `  files
```
np.save('tmp/x_train.npy', x_train)
np.save('tmp/y_train.npy', y_train)
np.save('tmp/x_val.npy', x_val)
np.save('tmp/y_val.npy', y_val)
```
thanks for your reply

**Comment by maxpumperla** [comment]
hmm, sorry. then this should work, unless the files aren't overwritten properly. I'll investigate more.

**Comment by Kirayue** [comment]
Hi, @maxpumperla 
I add `print('Data function called!')` to data function and I found that the data function was only called once at the first round.
![image](https://user-images.githubusercontent.com/13309212/41360757-15ed3860-6f60-11e8-90f7-1652b2eb7749.png)
I hope this might be helpful for you.

thanks for your reply

**Comment by xuzhang5788** [comment]
@Kirayue 
It is understandable that you used the different seeds, then you got the different l2 values for the best performance. However, in practical, which l2 will you use? What is the next actions? Thanks.
